In [16]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Conectamos las base de datos con python

In [17]:
query = "SELECT COUNT(*) FROM books;"
pd.read_sql(query, con=engine)

count
0   1000

Hice un query de prueba para verificar que se haya hecho la conexión correctamente

In [18]:
pd.read_sql("""
    SELECT table_name
    FROM information_schema.tables
    WHERE table_schema = 'public'
    ORDER BY table_name;
""", con=engine)

table_name
0  advertisment_costs
1             authors
2               books
3           check_avg
4              orders
5          publishers
6             ratings
7             reviews
8              visits

Verifique por medio del query las tablas con las que contaba 

In [19]:
# Tablas clave del proyecto
tablas = ['books', 'authors', 'publishers', 'ratings', 'reviews']

for tabla in tablas:
    print(f"\n📘 Explorando tabla: {tabla.upper()}")
    
    df = pd.read_sql(f"SELECT * FROM {tabla}", con=engine)
    
    print("\n🔹 Primeras filas:")
    display(df.head())
    
    print("\n🔹 Info general:")
    display(df.info())
    
    print("\n🔹 Nulos por columna:")
    display(df.isnull().sum())
    
    print("🔹 Duplicados:", df.duplicated().sum())


📘 Explorando tabla: BOOKS

🔹 Primeras filas:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


🔹 Info general:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None


🔹 Nulos por columna:


book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

🔹 Duplicados: 0

📘 Explorando tabla: AUTHORS

🔹 Primeras filas:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


🔹 Info general:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None


🔹 Nulos por columna:


author_id    0
author       0
dtype: int64

🔹 Duplicados: 0

📘 Explorando tabla: PUBLISHERS

🔹 Primeras filas:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


🔹 Info general:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None


🔹 Nulos por columna:


publisher_id    0
publisher       0
dtype: int64

🔹 Duplicados: 0

📘 Explorando tabla: RATINGS

🔹 Primeras filas:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


🔹 Info general:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None


🔹 Nulos por columna:


rating_id    0
book_id      0
username     0
rating       0
dtype: int64

🔹 Duplicados: 0

📘 Explorando tabla: REVIEWS

🔹 Primeras filas:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


🔹 Info general:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None


🔹 Nulos por columna:


review_id    0
book_id      0
username     0
text         0
dtype: int64

🔹 Duplicados: 0


Explore los Datasets para saber como estaban compuestos, si había nulos, los tipo de datos y los duplicados.

In [20]:
# leer las tablas como df

books_df = pd.read_sql("SELECT * FROM books", con=engine)
authors_df = pd.read_sql("SELECT * FROM authors", con=engine)
publishers_df = pd.read_sql("SELECT * FROM publishers", con=engine)
ratings_df = pd.read_sql("SELECT * FROM ratings", con=engine)
reviews_df = pd.read_sql("SELECT * FROM reviews", con=engine)

In [21]:
# Cambiar el tipo de en books de object a datetime

books_df['publication_date'] = pd.to_datetime(books_df['publication_date'])
display(books_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   book_id           1000 non-null   int64         
 1   author_id         1000 non-null   int64         
 2   title             1000 non-null   object        
 3   num_pages         1000 non-null   int64         
 4   publication_date  1000 non-null   datetime64[ns]
 5   publisher_id      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


None

hice el cambio de dtype en la columna publication_date de object a datetiem64

### Ejercicios

1. Encuentra el número de libros publicados después del 1 de enero de 2000.

In [36]:
query = """
SELECT COUNT(*) AS books_after_2000
FROM books
WHERE publication_date > '2000-01-01'
"""
pd.read_sql(query, con=engine)

books_after_2000
0               819

Se identificaron 819 libros publicados desde el año 2000.

2. Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [37]:
query = """
SELECT
    b.title AS book_title,
    COUNT(DISTINCT rv.review_id) AS total_text_reviews,
    COUNT(DISTINCT rt.rating_id) AS total_ratings,
    ROUND(AVG(rt.rating), 2) AS avg_rating
FROM books b
LEFT JOIN ratings rt ON b.book_id = rt.book_id
LEFT JOIN reviews rv ON b.book_id = rv.book_id
GROUP BY b.book_id, b.title
ORDER BY total_text_reviews DESC, total_ratings DESC
LIMIT 10
"""
book_engagement_summary = pd.read_sql(query, con=engine)
book_engagement_summary

book_title  total_text_reviews  \
0                            Twilight (Twilight  #1)                   7   
1                The Hobbit  or There and Back Again                   6   
2                             The Catcher in the Rye                   6   
3  Harry Potter and the Prisoner of Azkaban (Harr...                   6   
4  Harry Potter and the Chamber of Secrets (Harry...                   6   
5  The Lightning Thief (Percy Jackson and the Oly...                   6   
6             The Da Vinci Code (Robert Langdon  #2)                   6   
7                                      The Alchemist                   6   
8                          The Giver (The Giver  #1)                   6   
9                                     The Book Thief                   6   

   total_ratings  avg_rating  
0            160        3.66  
1             88        4.13  
2             86        3.83  
3             82        4.41  
4             80        4.29  
5             62        4.08  
6             59        3.83  
7             57        3.79  
8             56        3.75  
9             53        4.26

El libro con mayor número de calificaciones y reseñas es "Twilight", con 160 calificaciones y 7 reseñas de texto, y un promedio de 3.66. Le siguen clásicos como "The Hobbit" y títulos de Harry Potter.
Este análisis muestra qué libros despiertan más interacción cuantitativa (ratings) y cualitativa (texto), siendo un reflejo del compromiso del lector.

3. Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [38]:
query = """
SELECT
    p.publisher AS publisher_name,
    COUNT(b.book_id) AS total_books
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_books DESC
LIMIT 1
"""
top_publisher = pd.read_sql(query, con=engine)
top_publisher

publisher_name  total_books
0  Penguin Books           42

La editorial Penguin Books lidera en publicaciones de más de 50 páginas, con 42 títulos. Este resultado sugiere un catálogo sólido y amplio por parte de la editorial, orientado a obras más desarrolladas en contenido.

4. Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [39]:
query = """
SELECT
    a.author,
    ROUND(AVG(book_avg.avg_rating), 2) AS author_avg_rating
FROM (                                                                --Esta subconsulta calcula el promedio de calificaciones  para cada libro
    SELECT                                                             
        b.book_id,
        b.author_id,
        AVG(r.rating) AS avg_rating,
        COUNT(r.rating) AS rating_count
    FROM ratings r
    JOIN books b ON r.book_id = b.book_id
    GROUP BY b.book_id, b.author_id
    HAVING COUNT(r.rating) >= 50
) AS book_avg
JOIN authors a ON book_avg.author_id = a.author_id                  --Aquí agrupamos por autor, y sacamos su promedio entre sus libros aptos,
GROUP BY a.author                                                   -- es decir, con ≥ 50 ratings
ORDER BY author_avg_rating DESC
LIMIT 1
"""
top_author = pd.read_sql(query, con=engine)
top_author

author  author_avg_rating
0  J.K. Rowling/Mary GrandPré               4.28

El autor mejor posicionado fue J.K. Rowling/Mary GrandPré, con una calificación promedio de 4.28 en sus libros más populares. Esto refuerza su consistencia en la calidad percibida por los lectores, y la relevancia continua de su obra.

5. Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [41]:
query = """
SELECT
    ROUND(AVG(user_review_counts.total_reviews), 2) AS avg_text_reviews_per_user
FROM (
    SELECT
        rv.username,
        COUNT(rv.review_id) AS total_reviews
    FROM reviews rv
    WHERE rv.username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(*) > 50
    )
    GROUP BY rv.username
) AS user_review_counts
"""
avg_reviews = pd.read_sql(query, con=engine)
avg_reviews

avg_text_reviews_per_user
0                      24.33

Los usuarios que calificaron más de 50 libros escribieron en promedio 24.33 reseñas de texto. Esto demuestra que los usuarios más activos no solo califican, sino que también participan de forma cualitativa, aportando comentarios y reflexiones que enriquecen la comunidad.